lxd/main_init.go

```diff
@@ -494,8 +510,148 @@ they otherwise would.
 			return err
 		}
 	}
+	return nil
+}
+
+// Run the logic for preseed mode
+func (cmd *CmdInit) runPreseed(client lxd.ContainerServer) error {
+	data := &cmdInitData{}
+
+	err := cmd.Context.InputYAML(data)
+	if err != nil {
+		return fmt.Errorf("Invalid preseed YAML content")
+	}
+
+	return cmd.run(client, data)
+}
+
+// Apply the configuration specified in the given init data.
+func (cmd *CmdInit) run(client lxd.ContainerServer, data *cmdInitData) error {
```

Comment:
Ideally this function would be transactional but since that not possible we should at least try to revert to the previous state as best as we can. It can simply be similar to what we do in the `Update()` method in `lxd/container_lxc.go`:
- store the previous state of things (changes, newly created entities and so on)
- define a boolean `tryRevert` or something and then use a closure or a bunch of defers that only fire when `tryRevert` is true and do nothing when `tryRevert` is set to false at the end of the function.